File: src/discord/fileDownload.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import type { OutgoingFile } from './ports.js';

// Read-only file browse/download, realpath-confined to the session workspace
// (§4, §7.5, fixes A5). Every browsed/downloaded path is realpath-resolved and
// must stay inside the workspace root — a symlink pointing outside the root is
// caught, not just a literal `..`. Reuses the confinement approach from
// mcpFileTool/sessionOrchestrator (realpath the deepest existing ancestor).
//
// This is a "read" component: it lists entries and resolves a download path, but
// never mutates the filesystem. The actual Discord attachment is built by 7b from
// the OutgoingFile this returns. Pure FS logic — unit-testable with temp dirs.

export class WorkspaceEscapeError extends Error {
  constructor(public readonly requestedPath: string) {
    super(`Path escapes the workspace: ${requestedPath}`);
    this.name = 'WorkspaceEscapeError';
  }
}

export interface DirEntry {
  name: string;
  isDirectory: boolean;
}

export class FileDownload {
  private readonly root: string;

  constructor(workspaceRoot: string) {
    this.root = realpathOrResolve(workspaceRoot);
  }

  // List the entries of a directory inside the workspace. `relativeDir` is resolved
  // against the workspace root; anything escaping the root throws WorkspaceEscapeError.
  browse(relativeDir = '.'): DirEntry[] {
    const target = this.resolveConfined(relativeDir);
    const entries = fs.readdirSync(target, { withFileTypes: true });
    return entries
      .filter((e) => !e.name.startsWith('.'))
      .map((e) => ({ name: e.name, isDirectory: e.isDirectory() }))
      .sort((a, b) => {
        if (a.isDirectory !== b.isDirectory) return a.isDirectory ? -1 : 1;
        return a.name.localeCompare(b.name);
      });
  }

  // Resolve a download target inside the workspace, returning an OutgoingFile that
  // 7b turns into a discord.js attachment. Throws WorkspaceEscapeError when the path
  // escapes the workspace, or a plain Error when the target is missing / not a file.
  download(relativePath: string): OutgoingFile {
    const target = this.resolveConfined(relativePath);
    let stat: fs.Stats;
    try {
      stat = fs.statSync(target);
    } catch {
      throw new Error(`File not found: ${relativePath}`);
    }
    if (!stat.isFile()) {
      throw new Error(`Not a file: ${relativePath}`);
    }
    return { path: target, name: path.basename(target) };
  }

  // Resolve `requested` against the workspace root, realpath-confining it. Throws
  // WorkspaceEscapeError if the resolved path is outside the root.
  private resolveConfined(requested: string): string {
    const resolved = realpathOrResolve(path.resolve(this.root, requested));
    if (!isWithin(this.root, resolved)) {
      throw new WorkspaceEscapeError(requested);
    }
    return resolved;
  }
}

// Realpath a path, falling back to the realpath of its deepest existing ancestor
// joined with the non-existent tail — so confinement holds for paths that do not
// exist yet while still resolving symlinks in the part that does. (Same approach as
// core/sessionOrchestrator realpathOrResolve; duplicated locally to keep the
// discord layer free of a core import for a tiny helper.)
function realpathOrResolve(target: string): string {
  const abs = path.resolve(target);
  let existing = abs;
  const tail: string[] = [];
  while (!fs.existsSync(existing)) {
    const parent = path.dirname(existing);
    if (parent === existing) break;
    tail.unshift(path.basename(existing));
    existing = parent;
  }
  try {
    const realExisting = fs.realpathSync(existing);
    return tail.length > 0 ? path.join(realExisting, ...tail) : realExisting;
  } catch {
    return abs;
  }
}

function isWithin(root: string, child: string): boolean {
  const rel = path.relative(root, child);
  return rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel));
}

File: swift/Sources/DiscordAgentBridge/Session/FileDownload.swift
import Foundation

// Read-only file browse/download, realpath-confined to the session workspace
// (TS `fileDownload.ts`). Every browsed/downloaded path is realpath-resolved and
// must stay inside the workspace root — a symlink pointing outside the root is
// caught, not just a literal `..`. Pure FS logic — unit-testable with temp dirs.
//
// No Discord UI wiring (TS also leaves this as a library only). Future slash/UI
// can call `browse` / `download` and turn `OutgoingFile` into an attachment.

/// Thrown when a browse/download path escapes the workspace root.
public struct WorkspaceEscapeError: Error, CustomStringConvertible, Equatable {
    public let requestedPath: String

    public init(requestedPath: String) {
        self.requestedPath = requestedPath
    }

    public var description: String {
        "Path escapes the workspace: \(requestedPath)"
    }
}

/// One directory listing entry (TS `DirEntry`).
public struct DirEntry: Sendable, Equatable {
    public var name: String
    public var isDirectory: Bool

    public init(name: String, isDirectory: Bool) {
        self.name = name
        self.isDirectory = isDirectory
    }
}

/// Path + display name for a Discord attachment (TS `OutgoingFile` path form).
public struct OutgoingFile: Sendable, Equatable {
    public var path: String
    public var name: String

    public init(path: String, name: String) {
        self.path = path
        self.name = name
    }
}

/// Confined workspace browser/downloader (TS `FileDownload` class).
public final class FileDownload: @unchecked Sendable {
    private let root: String

    public init(workspaceRoot: String) {
        self.root = realpathOrResolve(workspaceRoot)
    }

    /// List entries of a directory inside the workspace. Dotfiles hidden; dirs first, then name.
    /// Throws `WorkspaceEscapeError` when `relativeDir` escapes the root.
    public func browse(relativeDir: String = ".") throws -> [DirEntry] {
        let target = try resolveConfined(relativeDir)
        let fm = FileManager.default
        let names = try fm.contentsOfDirectory(atPath: target)
        var entries: [DirEntry] = []
        for name in names where !name.hasPrefix(".") {
            let full = (target as NSString).appendingPathComponent(name)
            var isDir: ObjCBool = false
            guard fm.fileExists(atPath: full, isDirectory: &isDir) else { continue }
            entries.append(DirEntry(name: name, isDirectory: isDir.boolValue))
        }
        entries.sort { a, b in
            if a.isDirectory != b.isDirectory { return a.isDirectory && !b.isDirectory }
            return a.name.localizedStandardCompare(b.name) == .orderedAscending
        }
        return entries
    }

    /// Resolve a download target inside the workspace. Throws `WorkspaceEscapeError` on escape,
    /// or a plain `FileDownloadError` when missing / not a file.
    public func download(relativePath: String) throws -> OutgoingFile {
        let target = try resolveConfined(relativePath)
        var isDir: ObjCBool = false
        guard FileManager.default.fileExists(atPath: target, isDirectory: &isDir) else {
            throw FileDownloadError.notFound(relativePath)
        }
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: target),
              (attributes[.type] as? FileAttributeType) == .typeRegular
        else {
            throw FileDownloadError.notAFile(relativePath)
        }
        return OutgoingFile(path: target, name: (target as NSString).lastPathComponent)
    }

    private func resolveConfined(_ requested: String) throws -> String {
        let joined: String
        if (requested as NSString).isAbsolutePath {
            joined = requested
        } else {
            joined = (root as NSString).appendingPathComponent(requested)
        }
        let resolved = realpathOrResolve(joined)
        guard isWithin(root: root, child: resolved) else {
            throw WorkspaceEscapeError(requestedPath: requested)
        }
        return resolved
    }
}

public enum FileDownloadError: Error, CustomStringConvertible, Equatable {
    case notFound(String)
    case notAFile(String)

    public var description: String {
        switch self {
        case .notFound(let p): return "File not found: \(p)"
        case .notAFile(let p): return "Not a file: \(p)"
        }
    }
}
